### PR TITLE
[WIP] refactor initialization

### PR DESCRIPTION
**Pull request by therealplato**
We are currently loading all configuration in the init function of an env package.

This PR will refactor initialization to cleanly configure all broadway behavior from the entry point.

- [x] Create configuration struct
- [x] Populate configuration struct from environment variables
- [x] Pass configuration struct to dependencies
- [X] Update `server` package to use config instead of `env`
- [ ] Update `instance` package to use config instead of `env`
- [x] Update `deployment` package to use config instead of `env`
- [x] Update `notification` package to use config instead of `env`
- [ ] Update `services` package to use config instead of `env`
- [ ] Update `store` package to use config instead of `env`
- [X] Update `slack` package to use config instead of `env`
- [ ] Remove `env` package
- [x] Create cmd/bwserver/main.go
- [x] Populate configuration struct from command line flags
- [x] Start server with `go run cmd/bwserver/main.go` configured by env/flags
- [x] Vendor codedgangsta/cli #111 (unblocked)
- [x] <strike>Create cmd/bwctl/main.go</strike>
- [x] <strike>Support `go run cmd/bwctl/main.go ...` passing command line parameters to slack command parser</strike>  
`bwctl` binary will be a separate PR

### CODE REVIEW

File: cmd/server.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"fmt"
+
+	"github.com/namely/broadway/env"
+	"github.com/namely/broadway/server"
+	"github.com/namely/broadway/store"
+	"gopkg.in/urfave/cli.v1"
+)
+
+var ServerCmd = func(c *cli.Context) error {
```
Comment (therealplato) [inline cmd/server.go]: I KNOW DAWG

File: deployment/manifest.go [hunk 2]
```diff
@@ -45,30 +43,11 @@ func NewManifest(id, content string) (*Manifest, error) {
 // Execute executes template with variables
 func (m *Manifest) Execute(vars map[string]string) string {
 	var b bytes.Buffer
+	fmt.Printf("pre-executing %s...", m.ID)
```
Comment (therealplato) [inline deployment/manifest.go]: :bomb:

File: deployment/playbook.go [hunk 3]
```diff
@@ -132,3 +140,23 @@ func LoadPlaybookFolder(dir string) (map[string]*Playbook, error) {
 	}
 	return playbooks, nil
 }
+
```
Comment (therealplato) [inline deployment/playbook.go]: put back in deployment/manifest.go

File: server/server_test.go [hunk 4]
```diff
@@ -20,89 +20,64 @@ import (
...
+}
 
-func makeRequest(req *http.Request, w *httptest.ResponseRecorder) {
-	mem := etcdstore.New()
-
-	server := New(mem)
-	server.Init()
-	server.Handler().ServeHTTP(w, req)
+func makeRequest(s *Server, req *http.Request, w *httptest.ResponseRecorder) {
+	// mem := etcdstore.New()
```
Comment (therealplato) [inline server/server_test.go]: :bomb:

File: server/server_test.go [hunk 5]
```diff
@@ -125,40 +101,43 @@ func TestCreateInstanceWithInvalidAttributes(t *testing.T) {
...
 	_, err := service.CreateOrUpdate(i)
 	if err != nil {
 		t.Log(err.Error())
 	}
 
 	req, w := testutils.GetRequest(t, "/instance/helloplaybook/TestGetInstanceWithValidPath")
-	req = auth(req)
-	makeRequest(req, w)
+	req = auth(testCfg, req)
+	server := New(store, testCfg)
```
Comment (therealplato) [inline server/server_test.go]: todo: flip this function signature to be consistent with all other config injections

File: cfg/cfg.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	K8sCAFile          string // the CA file setting for local development
+	EtcdEndpoints      string // the list Etcd hosts separated by comma
+	EtcdPath           string // the root directory for Broadway objects
+	PlaybooksPath      string // the folder where playbooks are found
+	ManifestsPath      string // the folder where manifests are found
+	ManifestsExtension string // .yml or .yaml
+	AuthBearerToken    string // a global token required for all requests except GET/POST command/
+	SlackToken         string // the expected Slack custom command token.
+	ServerHost         string // passed to gin and configures the listen address of the server
+	SlackWebhook       string // your team's slack incoming message webhook URL
```
Comment (poopoothegorilla) [inline cfg/cfg.go]: @therealplato did you come up with this?

File: cmd/common.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		EnvVar:      "BROADWAY_MANIFESTS_PATH",
+		Destination: &cfg.GlobalCfg.ManifestsPath,
+	},
+	cli.StringFlag{
+		Name:        "manifest-extension",
+		Usage:       "the extension used by manifest files",
+		Value:       ".yml",
+		EnvVar:      "BROADWAY_MANIFESTS_EXTENSION",
+		Destination: &cfg.GlobalCfg.ManifestsExtension,
+	},
```
Comment (poopoothegorilla) [inline cmd/common.go]: 👍 looks so nice and clean!!!

File: cmd/server.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+// ServerCmd is executed by cli on `broadway server`
+var ServerCmd = func(c *cli.Context) error {
+	etcdstore.Setup(cfg.GlobalCfg)  // configure etcd before using
+	deployment.Setup(cfg.GlobalCfg) // configure kubernetes deployments before using
+	fmt.Printf("starting server with config...\n%+v", cfg.GlobalCfg)
+	s := server.New(etcdstore.New(), cfg.GlobalCfg)
+	s.Init()
+	err := s.Run(cfg.GlobalCfg.ServerHost)
+	if err != nil {
```
Comment (poopoothegorilla) [inline cmd/server.go]: scope

**Comment by poopoothegorilla** [comment]
This is really big! is it possible to split it up into some smaller chunks? or does it need to be together?

**Comment by therealplato** [comment]
> is it possible to split it up

oh i hope so that would be great. maybe i can do one PR per package
i had no idea how big it would get when I started :)
and 4/5 is vendored cli

File: cfg/cfg.go (hunk 6 above)
Comment (therealplato) [inline cfg/cfg.go]: this is largely a cleanup of [`env/env.go`](https://github.com/namely/broadway/blob/880c553b6054936173a3606acc10445ecebcb581/env/env.go#L11)

**Comment by therealplato** [comment]
Moving this to #121

**Pull request closed without merging**